Update domain auto linking

jojomatik opened the pull request:
<!--
Thanks for contributing to the project!
Please help us keep this project in good shape by going through this checklist.
Replace the empty checkboxes [ ] below with checked ones [X] as they are completed
Remember, you can preview this before saving it.
-->

<!-- You can remove this first section if you have contributed before -->

### First time contributor checklist:

- [x] I have read the [README](https://github.com/signalapp/Signal-Desktop/blob/master/README.md) and [Contributor Guidelines](https://github.com/signalapp/Signal-Desktop/blob/master/CONTRIBUTING.md)
- [x] I have signed the [Contributor Licence Agreement](https://signal.org/cla/)

### Contributor checklist:

- [x] My contribution is **not** related to translations. _Please submit translation changes via our [Signal Desktop Transifex project](https://www.transifex.com/signalapp/signal-desktop/)._
- [x] My commits are in nice logical chunks with [good commit messages](http://chris.beams.io/posts/git-commit/)
- [x] My changes are [rebased](https://medium.freecodecamp.org/git-rebase-and-the-golden-rule-explained-70715eccc372) on the latest [`development`](https://github.com/signalapp/Signal-Desktop/tree/development) branch
- [ ] A `yarn ready` run passes successfully ([more about tests here](https://github.com/signalapp/Signal-Desktop/blob/master/CONTRIBUTING.md#tests)) [fails at step `lint-deps`, I'm using [this workaround](https://community.signalusers.org/t/issues-compiling-signal-desktop-on-windows-environment/29332/4), as I'm having problems to build the app]
- [x] My changes are ready to be shipped to users

### Description

<!--
Describe briefly what your pull request changes. Focus on the value provided to users.

Does it address any outstanding issues in this project?
  https://github.com/signalapp/Signal-Desktop/issues?utf8=%E2%9C%93&q=is%3Aissue
  Reference an issue with the hash symbol: "#222"
  If you're fixing it, use something like "Fixes #222"

Please write a summary of your test approach:
  - What kind of manual testing did you do?
  - Did you write any new tests?
  - What operating systems did you test with? (please use specific versions: http://whatsmyos.com/)
  - What other devices did you test with? (other Desktop devices, Android, Android Simulator, iOS, iOS Simulator)
-->
#### Changes
Updated list of auto-linked TLDs (fixes #4538) in [Linkify.tsx](ts/components/conversation/Linkify.tsx)
- The new list features all TLDs registered with IANA in 2010 or earlier.
- See also:
  - relevant GitHub comment: #4538 (comment)
  - IANA root zone database: https://www.iana.org/domains/root/db
 
##### Before:
![image](https://user-images.githubusercontent.com/66277921/114307030-2212b280-9ade-11eb-8547-9df42138f018.png)

##### After:
![image](https://user-images.githubusercontent.com/66277921/114307037-28a12a00-9ade-11eb-8611-a62fbcd4cc56.png)
 
#### Documentation and testing
- Added stories for non-ASCII and RTL TLDs to [Linkify.stories.tsx](ts/components/conversation/Linkify.stories.tsx)
  - Added a story with a non-ASCII TLD (in this case hangul).
  - Added two stories with Right-to-left TLDs:
    - RTL only
    - RTL TLD mixed with Left-to-right text.
  
 - Manual visual testing in production version (v1.40.1) and my local development version on Windows 10 64bit (10.0.19041). For a comparison see sreenshots above.
 
#### Next steps?
As mentioned in [this comment](https://github.com/signalapp/Signal-Desktop/issues/4538#issuecomment-817292461) there are still two open questions:
1. While adding the new stories to the storybook I used an RTL domain and another non-ASCII domain (in this case an arabic and on in hangul as pointed out in [your comment](https://github.com/signalapp/Signal-Desktop/issues/4538#issuecomment-816882227)). As I don't speak korean, arabic or any other language in non latin scripts, I made up some domains. Should I rather use a domain that is in use or one that isn't? I don't want to endorse any content that I don't understand. 
2. Some of the domains (e.g. 테스트) in your list are marked as `Test top-level domain` in IANA's root zone database. Those would be great for testing purposes indeed, as they are unused, but I feel like those shouldn't be included in the production build. Should I update the list to remove those?

EvanHahn-Signal commented on ts/components/conversation/Linkify.tsx:
```diff
@@ -9,31 +9,314 @@ import { RenderTextCallbackType } from '../../types/Util';
-  //   this only applies to "fuzzy" matches (`example.com`), not matches with
-  //   protocols (`https://example.com`).
-  // [0]: https://en.wikipedia.org/wiki/Generic_top-level_domain#History
+  // This is all TLDs in place in 2010, according to [IANA's root zone database][0].
```
Given that we're planning to remove test TLDs, it could be worth calling that out here.

jojomatik replied:
You're right, I mentioned it in the new version.

EvanHahn-Signal commented on ts/components/conversation/Linkify.stories.tsx:
```diff
@@ -57,6 +57,32 @@ story.add('Missing Protocol', () => {
+story.add('Right-to-left TLD', () => {
+  const props = createProps({
+    text:
+      'أنا مرتبك قليلاً حول كيفية عمل النص من اليمين إلى اليسار ، ولكن هنا مجال: تونس.تونس', // I am slightly confused how RTL text works, but here is a domain:
```
Nitpick:

```suggestion
      // "I am slightly confused how RTL text works, but here is a domain:"
      'أنا مرتبك قليلاً حول كيفية عمل النص من اليمين إلى اليسار ، ولكن هنا مجال: تونس.تونس', 
```

jojomatik replied:
Probably correct in newer version.

EvanHahn-Signal reviewed (changes requested):
Thanks so much for doing this!

To answer your questions:

> 1. While adding the new stories to the storybook I used an RTL domain and another non-ASCII domain (in this case an arabic and on in hangul as pointed out in your comment). As I don't speak korean, arabic or any other language in non latin scripts, I made up some domains. Should I rather use a domain that is in use or one that isn't? I don't want to endorse any content that I don't understand.

Good question. I'd see if there are any analogs to `example.com` for these TLDs, and if you spend more than ~10 minutes and fail, use real URL. I'd try to find a URL that's "mainstream" (i.e., a government website or well-known company) in an effort to avoid any controversy. Thanks for considering this.

> 2. Some of the domains (e.g. 테스트) in your list are marked as Test top-level domain in IANA's root zone database. Those would be great for testing purposes indeed, as they are unused, but I feel like those shouldn't be included in the production build. Should I update the list to remove those?

Great catch. I don't know too much about these domains, but I see no reason to include them in any build. I was wrong to include them in my list.

jojomatik commented:
I think we could go with the domains `кц.рф` of the "Coordination Center for TLD RU" and `www.تجهيز.تونس` of the tunesian "Ministry of equipment, housing and infrastructure". What do you think?

*Most institutions seem to prefer the domains in latin script even though it's not their original script. It is actually kind of hard to find these without speaking the language and with an european google account. The russion one luckily had both domains in their SSL certificate.*

Should I rather edit my previous commits and force push or should I add another commit?

EvanHahn-Signal commented:
> Should I rather edit my previous commits and force push or should I add another commit?

Whatever's easier for you. We will squash-merge your changes, so the individual commits aren't too important

If you haven't, I also encourage you to rebase against or merge `development`, as we just did a release. Not a requirement, but it'll help make sure your branch is using up-to-date code.

jojomatik commented:
> If you haven't, I also encourage you to rebase against or merge `development`, as we just did a release. Not a requirement, but it'll help make sure your branch is using up-to-date code.

For me `development` had the last commit 4 days ago. Is it possible that it has not been updated to have the same state as `master` yet?

EvanHahn-Signal commented:
Ah, sorry—please ignore my comment.

jojomatik commented:
Hi there, I've force pushed the current state. Although I'm unsure if it's entirely correct.

There is (maybe) one problem with (mixed LTR-) RTL domains (e.g. `www.تجهيز.تونس`):

- it renders correctly in both examples in the storybook and the link works flawlessly.
- it doesn't render if I copy the string out of my code or the storybook into my Signal build.

Actually I suppose this is not an issue with the new code but an issue with me pasting the string into Signal or with Signal itself handling mixed RTL-LTR content, but I'm simply not sure. Also it doesn't make to much sense to me that the storybook would handle it differently than the running app. It's the same code after all, isn't it?

**Also:**
Theoretically everything would work without `www`, but something is off there as well (but **probably outside of Signal's scope**):
- Chrome shows an error `DNS_PROBE_FINISHED_NXDOMAIN` if i click on the link
- Firefox redirects me to `https://www.تجهيز.تونس/` which then reports error `PR_END_OF_FILE_ERROR` (chrome also reports an error when I try to use https, so probably the webserver configuration is off)
- If I take the complete working URL `http://www.تجهيز.تونس/index.php?id=2&L=` in chrome and strip everything off except the arabic characters it magically works.

Something might be off with the implementation of the mixed character set or the transcriped latin domain equivalent `xn--pgbes7fp.xn--pgbs0dh` being handled differently, but it might as well just me being illiterate in these scripts and making a stupid mistake. Or its partially the fault of the webserver, I'm not sure at all.

EvanHahn-Signal commented:
> Actually I suppose this is not an issue with the new code but an issue with me pasting the string into Signal or with Signal itself handling mixed RTL-LTR content, but I'm simply not sure. Also it doesn't make to much sense to me that the storybook would handle it differently than the running app. It's the same code after all, isn't it?

It should be the same code, but a few things could account for the difference:

1. The copy-paste, as you say, could change something
1. The "browser" running inside of Signal Desktop may not be identical to the browser you're using with Storybook. For example, if Signal Desktop uses Chrome 89 and you're hitting Storybook with Chrome 90, that could account for a difference.

You could debug this by hacking `<MessageBody>` to always use the same text, to see if it's a copy-paste issue or something else.

I don't think you need to do any of that, though, unless you want to. Linkification of these things is best-effort.

When you're ready, feel free to move this out of Draft mode and I'll kick the tires on it.

jojomatik commented:
> I don't think you need to do any of that, though, unless you want to. Linkification of these things is best-effort.

Yeah, if there actually was an issue (which might not even be the case) it would probably be really useful to have somebody literate in one of these RTL-scripts (probably the case if the issue arises) to take a look and experiment with it. 

I think from my point of view, it's ready then :)

hiqua commented:
@jojomatik while you're at it, you could also test domains such as the ones mentioned in this issue:
https://github.com/signalapp/Signal-Android/issues/11247

https://xn--kq9h.st/

jojomatik commented:
> @jojomatik while you're at it, you could also test domains such as the ones mentioned in this issue:
> [signalapp/Signal-Android#11247](https://github.com/signalapp/Signal-Android/issues/11247)

Yes, I'll have a look. 

If I find some time, I might also try to resolve signalapp/Signal-Android#11247. As I'm more proficient in java compared to ts/ js, that should actually be easier for me :)

jojomatik commented:
Just took a look at it. In [`link_previews.js` (line 156)](https://github.com/signalapp/Signal-Desktop/blob/dfbbd16e509ec1c2cccb6447a0c74855b1c8149b/js/modules/link_previews.js#L156) there is similar code to what is mentioned in signalapp/Signal-Android#11247. It specifically forbids domains consisting of ASCII and non-ASCII characters. 

I don't know why that is exactly, as - as mentioned in https://github.com/signalapp/Signal-Desktop/pull/5170#issuecomment-820635420 - there are public institutions using these mixed domains and it is allowed by the competent domain authorities.

The first version of this code was introduced in https://github.com/signalapp/Signal-Desktop/commit/ae2db9f09ac5de9356ee47f02eb629e35e370d47 and then simplified to the current state in https://github.com/signalapp/Signal-Desktop/commit/f5be32ba14a2f604f2016a75c2155741d6e7d446. 

My proposal would be to remove the following section:
```js
if (isMixed) {
  return true;
}
```
This would also allow to remove the 11 lines above, the `ASCII_PATTERN` constant and the import statement `const nodeUrl = require('url');`.

Another question would be why the mixed domain (i.e. `www.تجهيز.تونس`) mentioned in https://github.com/signalapp/Signal-Desktop/pull/5170#issuecomment-820635420 was rendered as a link in the storybook. Is it somehow ignoring this check?

scottnonnenberg-signal commented:
@jojomatik We use that mixed domains check as a rough heuristic for homograph attacks (for example, using a cyrillic character that looks exactly like a normal ASCII character). If you can provide more examples of valid mixed domains, that would be helpful.

jojomatik commented:
> @jojomatik We use that mixed domains check as a rough heuristic for homograph attacks (for example, using a cyrillic character that looks exactly like a normal ASCII character). If you can provide more examples of valid mixed domains, that would be helpful.

Ok, that is actually fairly important. Although I think it might be safe (enough) to allow mixing in total but to require to stay consistent at each level of the domain (such as `non-ASCII.ASCII.non-ASCII` for example). I would think that mixing at one level is not allowed by IANA anyway but I'm not sure about that. If this is not an IANA-requirement anyway I think it should be a criterion for linking (i think the old code from https://github.com/signalapp/Signal-Desktop/commit/ae2db9f09ac5de9356ee47f02eb629e35e370d47 already did something similar, but with hardcoded character sets; ASCII and non-ASCII should suffice in my opinion).

I have no statistics on mixed domain names, but [HKIRK (cf. p. 2)](https://www.hkirc.hk/upload/page/38/pdf_1/5e6b172e68b29.pdf)  for example allows the `.hk` variant with latin category domains, the `.hk` variant with chinese category names and the `.香港` variant with chinese category names. I suppose similar things are possible with other NICs. I don't think mixed domains are all that popular though, with maybe the exception of a `www.` subdomain.

jojomatik commented:
I did some more research on IDN and homograph attacks and found some interesting resources:
- [Blog post on dev.to](https://dev.to/logan/homographs-attack--5a1p)
- [Google Chrome's IDN policy](https://chromium.googlesource.com/chromium/src/+/main/docs/idn.md#google-chrome_s-idn-policy)
- [a relevant unicode technical standard](http://unicode.org/reports/tr39/#Whole_Script_Confusables)

I've not yet built a final opinion how I would like to handle it (in the end it's your opinion that matters anyway), but I feel like chromes policy is a bit to elaborate. It elimates the count of false positives, but I think a simpler algorithm (such as checking if each non-ASCII character in a block has an ASCII look-alike, based on a static list; this would allow some/ most non-ASCII/ mixed domains to render and would only block those that are entirely homographic) should suffice. Personally I don't give to much weight to these false positives. 

There are also some more recent papers using OCR which is really cool, but not useful here (if someone is interested though i could add some links to these).

hiqua commented:
@jojomatik are you still working on this or are you just waiting for a review from the Signal devs? It's not clear from your last comment. I'd advise against yak shaving here, if the current PR is already a clear improvement, then maybe try to get it merged before it becomes stale.

jojomatik commented:
> @jojomatik are you still working on this or are you just waiting for a review from the Signal devs? It's not clear from your last comment. I'd advise against yak shaving here, if the current PR is already a clear improvement, then maybe try to get it merged before it becomes stale.

That's true, I was not really clear. Theoretically this PR is still ready for review/ merging if the following question would be answered/ somebody would have a look at this:
> Another question would be why the mixed domain (i.e. www.تجهيز.تونس) mentioned in #5170 (comment) was rendered as a link in the storybook. Is it somehow ignoring this check?

I might have to adjust the example depending on the answer.

*I'd still be interested in implementing a good solution for signalapp/Signal-Android#11247 though and a corresponding one in this repo, but I'd need an opinion from the Signal devs on this. I think this should be consistent on Android and Desktop (and theoretically iOS, but I have no experience in iOS development). Also I feel like it would be more useful to have a separate PR for that.*

jojomatik commented:
I have thought about my question a bit more and found the difference between electron and chrome:

> Another question would be why the mixed domain (i.e. [www.تجهيز.تونس](http://www.%D8%AA%D8%AC%D9%87%D9%8A%D8%B2.%D8%AA%D9%88%D9%86%D8%B3)) mentioned in #5170 (comment) was rendered as a link in the storybook. Is it somehow ignoring this check?

https://github.com/signalapp/Signal-Desktop/blob/525525fd0c642976663f0c1aa0566d33b49256e8/js/modules/link_previews.js#L141

`nodeUrl.domainToUnicode` evaluates to `undefined` in my chrome version. And therefore the punified domain name is checked for non-ASCII characters and since the whole point of punified domain names is that they are ASCII-only the domain is not marked as sneaky and is therefore rendered in the storybook.

I have rebased on `development`, removed the `www.` from the tunesian domain as it should not work at least for now (and doesn't, at least inside Signal Desktop) and force pushed. Feel free to review and merge @EvanHahn-Signal or @scottnonnenberg-signal.

EvanHahn-Signal commented:
Thanks. We'll give this a review as soon as we can (though there's a lot on our plates right now, so it might not be soon).

jojomatik commented:
> Thanks. We'll give this a review as soon as we can (though there's a lot on our plates right now, so it might not be soon).

Perfect, take all the time you need. I don't think this PR is urgent by any means.

EvanHahn-Signal reviewed (approved):
Sorry to take so long here.

I'm going to pull this down, do a small amount of testing, change a few comments, and then merge.

Thanks for doing this and for your patience!

jojomatik commented:
Great to see the code slowly making its way into the software. Glad I could help! 🎉 😃
